### PR TITLE
feat: adds support for [sections] - GitLab style

### PR DESCRIPTION
**Pull request by sverweij**
## Description

- Adds support for sections as in use in GitLab 
  They're not illegal on GitHub (CODEOWNERS with them do check out OK), but I'm not sure
  they're actually supported yet.

## Motivation and Context

- They're 'legal' parts of a CODEOWNERS file both GitHub and on GitLab (and on GitLab they have a documented meaning). Before this PR the virtual-codeowners parser would most of the time reject lines with sections as 'unkown' syntax.
- Sections can contain group names that need to be expanded in CODEOWNERS

## How Has This Been Tested?

- [x] green ci
- [x] Additional unit/ integration tests & expanded _corpus_

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation only change
- [ ] Refactor (non-breaking change which fixes an issue without changing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] :book:

  - My change doesn't require a documentation update, or ...
  - it _does_ and I have updated it

- [x] :balance_scale:
  - The contribution will be subject to [The MIT license](https://github.com/sverweij/virtual-code-owners/blob/main/LICENSE), and I'm OK with that.
  - The contribution is my own original work.
  - I am ok with the stuff in [**CONTRIBUTING.md**](https://github.com/sverweij/virtual-code-owners/blob/main/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by McMeadow — COMMENTED** [review]
Good addition - you might want to look at `.github/CODEOWNERS` which still seems to contain test data, though

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -15,4 +15,17 @@
...
-README.md @sverweij
\ No newline at end of file
+README.md @sverweij
+
+[section]
+bla @sverweij
+
+[section-with-owners][(1)] @mcmeadow @sverweij
+# blabla # Patronen ohne Nutzernahmen sind erlaubt in Sectionen mit Nutzernamen im Titel
+bladiebla @other-owner-still
```
Comment (McMeadow) [inline .github/CODEOWNERS:25]: left over test data?

File: .github/labeler.yml [hunk 2]
```diff
@@ -12,15 +12,18 @@ vco:
   - changed-files:
       - any-glob-to-any-file: "**"
       - any-glob-to-any-file: README.md
+      - any-glob-to-any-file: bla
```
Comment (McMeadow) [inline .github/labeler.yml:15]: same here

File: .github/VIRTUAL-CODEOWNERS.txt [hunk 3]
```diff
@@ -7,4 +7,17 @@
...
-README.md @vco @documentation
\ No newline at end of file
+README.md @vco @documentation
+
+[section]
+bla @vco
+
+[section-with-owners] @vco @vco-admins
+# blabla # Patronen ohne Nutzernahmen sind erlaubt in Sectionen mit Nutzernamen im Titel
+bladiebla @other-owner-still
```
Comment (McMeadow) [inline .github/VIRTUAL-CODEOWNERS.txt:17]: might want to give this a look

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -15,4 +15,17 @@
...
 .github/  @mcmeadow @sverweij
 tools/    @mcmeadow @sverweij
-README.md @sverweij
\ No newline at end of file
+README.md @sverweij
+
+[section]
+bla @sverweij
+
+[section-with-owners][(1)] @mcmeadow @sverweij
```
Comment (McMeadow) [inline .github/CODEOWNERS:23]: `[(1)]` looks like erroneously generated. Generator seems ok though - leftover from a previous run?

File: .github/VIRTUAL-CODEOWNERS.txt (hunk 3 above)
Comment (sverweij) [reply .github/VIRTUAL-CODEOWNERS.txt]: check, fixed!

File: src/codeowners/generate.ts [hunk 5]
```diff
@@ -50,6 +50,23 @@ function generateLine(
       (pCSTLine.inlineComment ? ` #${pCSTLine.inlineComment}` : "")
     );
   }
+  if (pCSTLine.type === "section") {
+    const lUserNames = uniq(
+      pCSTLine.users.flatMap((pUser) => expandTeamToUserNames(pUser, pTeamMap)),
+    )
```
Comment (McMeadow) [inline src/codeowners/generate.ts:56]: this line is a duplicate. not per  se an issue if it's the clearest way to  express this. Is it indeed?

File: src/labeler-yml/generate.ts [hunk 6]
```diff
@@ -48,12 +48,12 @@ function getPatternsForTeam(
...
         return (
           pLine.type === "rule" &&
-          lineContainsTeamName(pLine as IInterestingCSTLine, pTeamName)
+          lineContainsTeamName(pLine as IRuleCSTLine, pTeamName)
         );
       })
       // @ts-expect-error ts thinks it can still be an IBoringCSTLine,
       // but with the filter above we've ruled that out
-      .map((pLine: IInterestingCSTLine) => pLine.filesPattern)
+      .map((pLine: IRuleCSTLine) => pLine.filesPattern)
```
Comment (McMeadow) [inline src/labeler-yml/generate.ts:56]: nice to  express that the labeller generator  currently only supports lines with _rules_ on them and not sections

File: src/virtual-code-owners/cst.d.ts [hunk 7]
```diff
@@ -14,6 +20,18 @@ export interface IInterestingCSTLine {
   inlineComment: string;
   raw: string;
 }
+export interface ISectionCSTLine {
+  type: "section";
```
Comment (McMeadow) [inline src/virtual-code-owners/cst.d.ts:24]: technically a _section head_ (see the  GitLab documentation), the section is the  head + the rules. Maybe rename while we can?

File: src/virtual-code-owners/parse.ts [hunk 8]
```diff
@@ -54,6 +58,49 @@ function parseLine(
...
+    sectionName: lSection.groups.sectionName as string,
+    spaces: lSection.groups.spaces as string,
+    users: parseUsers(lSection.groups.userNames as string, pTeamMap),
+    inlineComment: lTrimmedLine.split(/\s*#/)[1] ?? "",
+    raw: pUntreatedLine,
+  };
+
+  if (lSection.groups.minApprovers) {
+    lReturnValue.minApprovers = parseInt(
+      lSection.groups.minApprovers as string,
```
Comment (McMeadow) [inline src/virtual-code-owners/parse.ts:97]: radix?

File: src/virtual-code-owners/parse.ts [hunk 9]
```diff
@@ -54,6 +58,49 @@ function parseLine(
...
+    );
```
Comment (McMeadow) [inline src/virtual-code-owners/parse.ts:98]: Add base?
```suggestion
    lReturnValue.minApprovers = parseInt(
      lSection.groups.minApprovers as string,
      10
    );
```

File: src/virtual-code-owners/read.ts [hunk 10]
```diff
@@ -32,7 +32,7 @@ function reportAnomalies(pFileName: string, pAnomalies: IAnomaly[]): string {
   return pAnomalies
     .map((pAnomaly) => {
       if (pAnomaly.type === "invalid-line") {
-        return `${pFileName}:${pAnomaly.line}:1 invalid line - neither a rule, comment nor empty: "${pAnomaly.raw}"`;
+        return `${pFileName}:${pAnomaly.line}:1 invalid line - neither a rule, section, comment nor empty: "${pAnomaly.raw}"`;
```
Comment (McMeadow) [inline src/virtual-code-owners/read.ts:35]:
```suggestion
        return `${pFileName}:${pAnomaly.line}:1 invalid line - neither a rule, section head, comment nor empty: "${pAnomaly.raw}"`;
```

File: src/virtual-code-owners/cst.d.ts (hunk 7 above)
Comment (sverweij) [reply src/virtual-code-owners/cst.d.ts:24]: Good point. 

- [ ] Will update

File: src/codeowners/generate.ts (hunk 5 above)
Comment (sverweij) [reply src/codeowners/generate.ts]: looks like it's better slightly de-duplicated idd.

**Review by McMeadow — APPROVED** [review]
:shipit:

**Pull request merged**